praxis/layers/fp8_ops.py

```diff
@@ -0,0 +1,186 @@
+  return x.astype(dq_dtype) * scale.astype(dq_dtype)
+
+def quantize_dequantize(x, q_dtype, scale, compute_dtype):
+  qx = quantize(x, q_dtype, scale, compute_dtype)
+  return dequantize(qx, x.dtype, scale)
+
+def compute_scale(amax, scale, fp8_max, margin=0):
+  """Default function to convert amax to scaling factor."""
+  # This function copied from the TransformerEngine is used to compute its
+  # `scale`. However, our scale matches its `scale_inv` concept. So, we apply
+  # the reciprocal operation at the entry and exit of the function.
+  scale = 1.0 / scale
+  exp = jnp.floor(jnp.log2(fp8_max / amax)) - margin
+  sf = jnp.round(lax.pow(2., jnp.abs(exp)))
+  sf = jnp.where(amax > 0.0, sf, scale)
+  sf = jnp.where(lax.is_finite(amax), sf, scale)
+  sf = jnp.where(exp < 0, 1.0 / sf, sf)
+  return 1.0 / sf
+
+def compute_scale_and_amax_history(x, q_dtype, scale, amax_history):
+  dtype_max = get_fp8_max(q_dtype, jnp.float32)
+
+  amax_update = jnp.max(jnp.abs(x)).astype(scale.dtype)
+  new_amax_history = \
+      jnp.roll(amax_history, shift=-1, axis=0).at[0].set(amax_update)
+
+  amax_from_history = jnp.max(new_amax_history, axis=0)
+  new_scale = compute_scale(amax_from_history, scale, dtype_max)
+  return new_scale, new_amax_history
+
+def qdq_and_return(x, q_dtype, scale, amax_history, compute_dtype):
+  qx = quantize_dequantize(x, q_dtype, scale, compute_dtype)
+  new_scale, new_amax_history = compute_scale_and_amax_history(
+      x, q_dtype, scale, amax_history)
+  return qx, new_scale, new_amax_history
+
+@partial(custom_vjp, nondiff_argnums=(0,))
+def in_qdq(compute_dtype, inp, scale, amax_history):
+  qin, _, _ = qdq_and_return(
+      inp, jnp.float8_e4m3fn, scale, amax_history, compute_dtype)
+  return qin
+
+def in_qdq_fwd(compute_dtype, inp, scale, amax_history):
+  qin, new_scale, new_amax_history = qdq_and_return(
+      inp, jnp.float8_e4m3fn, scale, amax_history, compute_dtype)
+  return qin, (new_scale, new_amax_history)
+
+def in_qdq_bwd(compute_dtype, res, g):
+  new_scale, new_amax_history = res
+  q_g = g
+  return q_g, new_scale, new_amax_history
+
+in_qdq.defvjp(in_qdq_fwd, in_qdq_bwd)
+
+
+@partial(custom_vjp, nondiff_argnums=(0,))
+def out_qdq(compute_dtype, out, scale, amax_history):
+  return out
+
+def out_qdq_fwd(compute_dtype, out, scale, amax_history):
+  return out, (scale, amax_history)
+
+def out_qdq_bwd(compute_dtype, res, g):
+  scale, amax_history = res
+  q_g, new_scale, new_amax_history = qdq_and_return(
+      g, jnp.float8_e5m2, scale, amax_history, compute_dtype)
+  return q_g, new_scale, new_amax_history
+  
+out_qdq.defvjp(out_qdq_fwd, out_qdq_bwd)
+
+
+def fp8_einsum(eqn, x, k, compute_dtype, x_scale, x_amax_history, k_scale,
+               k_amax_history, dy_scale, dy_amax_history):
+  """Perform any einsum formula.  """
+
+  x_qdq = in_qdq(compute_dtype, x, x_scale, x_amax_history)
+
+  k_qdq = in_qdq(compute_dtype, k, k_scale, k_amax_history)
+
+  y_qdq = jnp.einsum(eqn, x_qdq, k_qdq)
```

Comment:
I am thinking this might need a significant effort to change the xla side for the pattern matching and specifically, we need to be careful about the scaling factors, since they are required for the fp8_gemm:
Basically, what we are doing now is like
```
high_precision_in1->QDQ1->einsum->high_precision_output
high_precision_in2->QDQ2->
```
We pattern match the above to:
```
high_precision_in1->Q1------>(DQ+gemm, which is fp8_gemm)->output
scale1(found in DQ1)--------->
high_precision_in2->Q2------>
scale2(found in DQ2)--------->
```
As for your suggestion, this is like what you want to do:
```
Q1->einsum->DQ1->DQ2
Q2->
```
These double epilog DQs are actually similar with our cudnn's solution (see DscaleQ and DscaleK in Figure 47 in [link](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#flash-fused-multi-head-att-fp8)). And if we want to pattern match this, we need to be careful about looking for the right scale for the right input. 

Also, I played with that `lax.dot(fp8_in1, fp8_in2)` example. And yes, I verified it is ok with the fp8 inputs and can use the fp8 gemm, but the scaling factors seem to be some default value (like 1.0, I guess) shared for all the input/output. So, at least, doing the QDQ in the suggested way is possible now since the dot/einsum is compatible with fp8 dtypes. But still as I mentioned above, the pattern matching might be rewritten in XLA.

cc @philipphack @wenscarl